gapis/resolve/dependencygraph2/dependency_graph.go

```diff
@@ -0,0 +1,425 @@
+	Index          int
+}
+
+func (ObsNode) dependencyNode() {}
+
+// Information about what sort of data to store in a dependency graph
+type DependencyGraphConfig struct {
+	// MergeSubCmdNodes indicates whether the graph should have one node per
+	// command (true), or a separate node for each subcommand (false)
+	MergeSubCmdNodes bool
+
+	// IncludeInitialCommands indicates whether nodes should be created for
+	// the initial (state rebuild) commands
+	IncludeInitialCommands bool
+}
+
+// NodeID identifies a node in a dependency graph
+type NodeID uint32
+
+const NodeNoID = NodeID(math.MaxUint32)
+
+// DependencyGraph stores the dependencies among api calls and memory observations,
+type DependencyGraph interface {
+
+	// NumNodes returns the number of nodes in the graph
+	NumNodes() int
+
+	// NumDependencies returns the number of dependencies (edges) in the graph
+	NumDependencies() int
+
+	// GetNode returns the node data associated with the given NodeID
+	GetNode(NodeID) Node
+
+	// GetNodeID returns the NodeID associated with given node data
+	GetNodeID(Node) NodeID
+
+	// ForeachCmd iterates over all API calls in the graph
+	ForeachCmd(ctx context.Context, cb func(context.Context, api.CmdID, api.Cmd) error) error
+
+	// ForeachNode iterates over all nodes in the graph
+	ForeachNode(cb func(NodeID, Node) error) error
+
+	// ForeachDependency iterates over all pairs (src, tgt), where src depends on tgt
+	ForeachDependency(cb func(NodeID, NodeID) error) error
+
+	// ForeachDependencyFrom iterates over all the nodes tgt, where src depends on tgt
+	ForeachDependencyFrom(src NodeID, cb func(NodeID) error) error
+
+	// ForeachDependencyTo iterates over all the nodes src, where src depends on tgt.
+	// If Config().ReverseDependencies is false, this will return an error.
+	ForeachDependencyTo(tgt NodeID, cb func(NodeID) error) error
+
+	// Capture returns the capture whose dependencies are stored in this graph
+	Capture() *capture.Capture
+
+	// GetCommand returns the command identified by the given CmdID
+	GetCommand(api.CmdID) api.Cmd
+
+	// InitialCommands returns the initial commands, which reconstruct the
+	// initial state before the first command in the capture.
+	InitialCommands() []api.Cmd
+
+	// InitialCmdRanges returns the memory ranges used by the initial commands.
+	InitialCmdRanges() interval.U64RangeList
+
+	// Config returns the config used to create this graph
+	Config() DependencyGraphConfig
+}
+
+type obsNodeIDs struct {
+	numReads int
+	nodeIDs  []NodeID
+}
+
+type dependencyGraph struct {
+	capture           *capture.Capture
+	cmdNodeIDs        *api.SubCmdIdxTrie
+	initMemoryNodeIDs obsNodeIDs
+	initCmdObsNodeIDs []obsNodeIDs
+	cmdObsNodeIDs     []obsNodeIDs
+	initialCommands   []api.Cmd
+	initialCmdRanges  interval.U64RangeList
+	nodes             []Node
+	numDependencies   int
+	dependenciesFrom  [][]NodeID
+	dependenciesTo    [][]NodeID
+
+	config DependencyGraphConfig
+}
+
+const initCmdBit = api.CmdID(1 << 62)
```

Comment:
Should we mention that this is the same to api/cmd_id.go: `derivedBit`? Or some comment why choose the same value, or why it does not matter to use the same value?

Reply:
The only reason it is redefined here is that `derivedBit` is not exported. Upon further reflection, this is a bad coupling between these packages; I will move the bit manipulations into `CmdID` methods.